### PR TITLE
Ticket 3089 (Halo node properties in Cylindrical2dMesh)

### LINKED ISSUE #142: Tidy up Bounded Cylindrical Voronoi Tesselations
Following on from trac Ticket https://chaste.cs.ox.ac.uk/trac/ticket/3078

We should look at the case of voids in Cylindrical meshes. This changes how we calculate the bounded voronoi tesselation.

In addition we want to make halo nodes parameters variable in Cylindrical2dMesh by adding new member variables 

mHaloSeparation
mHaloOffset

see https://chaste.cs.ox.ac.uk/trac/ticket/3089 for original ticket

### PR DESCRIPTION
**Pull request by jmosborne**
Add some member variables to change properties of Halo Nodes in Cylindrical2dMesh. #3089
Also change bounded cylindrical meshes to only have 1 ghost image node for each boundary edge, as more stable #3078

### CODE REVIEW

**Comment by jmosborne** [comment]
Need to update archive on a boost 1.55 machine. current archive is generated on 1.70 on docker.

**Comment by jmosborne** [comment]
Need to add test of generating a vertex mesh for a bounded VT mesh including where the image nodes are close to existing nodes to complete coverage

File: mesh/src/mutable/Cylindrical2dMesh.cpp [hunk 1]
```diff
@@ -160,10 +163,10 @@ void Cylindrical2dMesh::CreateHaloNodes()
     mTopHaloNodes.clear();
     mBottomHaloNodes.clear();
 
-    unsigned num_halo_nodes = (unsigned)(floor(mWidth*2.0));
+    unsigned num_halo_nodes = (unsigned)(floor(mWidth*mHaloScalingFactor));
```
Comment (fcooper8472) [inline mesh/src/mutable/Cylindrical2dMesh.cpp]: Ideally these old c-style casts should be replaced with `static_cast`, e.g.
```cpp
unsigned num_halo_nodes = static_cast<unsigned>(floor(mWidth*mHaloScalingFactor));
```

Comment (fcooper8472) [inline mesh/src/mutable/Cylindrical2dMesh.cpp]: Needs to have a formatter run over this: e.g. there should be a single space either side of binary operators such as * and /

The style is contained in the .clang-format file in the repo: all main IDEs will auto-format based on that style file.

File: mesh/src/mutable/Cylindrical2dMesh.cpp [hunk 2]
```diff
@@ -844,6 +847,28 @@ unsigned Cylindrical2dMesh::GetCorrespondingNodeIndex(unsigned nodeIndex)
     return corresponding_node_index;
 }
 
+double Cylindrical2dMesh::GetHaloScalingFactor()
```
Comment (fcooper8472) [inline mesh/src/mutable/Cylindrical2dMesh.cpp]: Method should be const

File: mesh/src/mutable/Cylindrical2dMesh.cpp [hunk 3]
```diff
@@ -844,6 +847,28 @@ unsigned Cylindrical2dMesh::GetCorrespondingNodeIndex(unsigned nodeIndex)
...
+    return mHaloScalingFactor;
+}
+
+void Cylindrical2dMesh::SetHaloScalingFactor(double haloScalingFactor)
+{
+    assert(haloScalingFactor >= 0.0);
+    mHaloScalingFactor = haloScalingFactor;
+}
+
+double Cylindrical2dMesh::GetHaloOffset()
```
Comment (fcooper8472) [inline mesh/src/mutable/Cylindrical2dMesh.cpp]: method should be const

File: mesh/src/vertex/Cylindrical2dVertexMesh.cpp [hunk 4]
```diff
@@ -150,40 +152,100 @@ Cylindrical2dVertexMesh::Cylindrical2dVertexMesh(Cylindrical2dMesh& rMesh, bool
...
-                        {
-                            double distance = norm_2(mpDelaunayMesh->GetVectorFromAtoB(nodes[i]->rGetLocation(), new_node_location));
-                            if (distance < node_clearance)
-                            {
-                                node_clear = false;
-                                //break;
-                            }
+                    
+                    double ratio = 0.5;
+                    c_vector<double,2> new_node_location = normal_vector + p_node_a->rGetLocation() + ratio*edge;
```
Comment (fcooper8472) [inline mesh/src/vertex/Cylindrical2dVertexMesh.cpp]: formatting: space either side of *

File: mesh/src/vertex/Cylindrical2dVertexMesh.cpp [hunk 5]
```diff
@@ -150,40 +152,100 @@ Cylindrical2dVertexMesh::Cylindrical2dVertexMesh(Cylindrical2dMesh& rMesh, bool
...
+                    
+                    // Check if near other nodes (TODO look at this as could be ineficient)
```
Comment (fcooper8472) [inline mesh/src/vertex/Cylindrical2dVertexMesh.cpp]: Is this todo still relevant?

File: mesh/src/vertex/Cylindrical2dVertexMesh.cpp [hunk 6]
```diff
@@ -150,40 +152,100 @@ Cylindrical2dVertexMesh::Cylindrical2dVertexMesh(Cylindrical2dMesh& rMesh, bool
...
+                    bool node_clear = true;
+                    double node_clearance = 0.01; 
+
+                    for (unsigned i=0; i<nodes.size(); i++)
+                    {   
+                        double distance = norm_2(mpDelaunayMesh->GetVectorFromAtoB(nodes[i]->rGetLocation(), new_node_location));
+                        if (distance < node_clearance)
+                        {   
+                            node_clear = false;
+                            //break;
```
Comment (fcooper8472) [inline mesh/src/vertex/Cylindrical2dVertexMesh.cpp]: If this break is not required it should be removed.

File: mesh/src/vertex/Cylindrical2dVertexMesh.cpp [hunk 7]
```diff
@@ -150,40 +152,100 @@ Cylindrical2dVertexMesh::Cylindrical2dVertexMesh(Cylindrical2dMesh& rMesh, bool
...
-                            nodes.push_back(new Node<2>(new_node_index, new_node_location));
-                            new_node_index++;
-                        }
+        // Add new nodes for voids (note these nodes won't be labeled as boundary nodes here)
+        for (TetrahedralMesh<2,2>::ElementIterator elem_iter = mpDelaunayMesh->GetElementIteratorBegin();
+            elem_iter != mpDelaunayMesh->GetElementIteratorEnd();
+            ++elem_iter)
+        {   
+            bool bad_element = false;
+            double edge_threshold = 1.5; //TODO Make setable variable!
```
Comment (fcooper8472) [inline mesh/src/vertex/Cylindrical2dVertexMesh.cpp]: Agreed, this looks like it should be settable!

File: mesh/src/vertex/Cylindrical2dVertexMesh.cpp [hunk 8]
```diff
@@ -364,7 +425,7 @@ void Cylindrical2dVertexMesh::CheckNodeLocation(Node<2>* pNode)
     {
         pNode->rGetModifiableLocation()[0] = x_location + mWidth;
     }
-    else if (x_location > mWidth)
+    else if (x_location >= mWidth)
```
Comment (fcooper8472) [inline mesh/src/vertex/Cylindrical2dVertexMesh.cpp:428]: Why has this changed?

File: mesh/test/mutable/TestCylindrical2dMesh.hpp [hunk 9]
```diff
@@ -325,6 +325,61 @@ class TestCylindrical2dMesh : public CxxTest::TestSuite
         TS_ASSERT_EQUALS(p_mesh->GetNumElements(), 2*cells_across*(cells_up-1));
         TS_ASSERT_EQUALS(p_mesh->GetNumBoundaryElements(), 1u); // boundary elements removed now halo nodes are used
     }
+    /*
```
Comment (fcooper8472) [inline mesh/test/mutable/TestCylindrical2dMesh.hpp:329]: formatting: if you run the formatter over these changes the whitespace will be fixed

**Review by fcooper8472 — CHANGES_REQUESTED** [review]
@jmosborne the code looks fine - there are a few todos that you might want to address, and some odd things like commented lines that perhaps should be removed.

Various parts of the code don't conform to our style guide. This is really simple to fix now, because we use clang format (see the .clang-format file in the repository). All modern IDEs will happily auto-format code blocks (or entire files) for you. Should be simple to set up: which IDE are you currently using?

Finally, this was opened in ticket 3089 on trac (in the archive here: https://github.com/Chaste/trac_archive/blob/master/issues/3089.md). Could you open a new issue please, so that it's possible to follow the PR through to a GitHub issue explaining the purpose of the change?

File: mesh/src/mutable/Cylindrical2dMesh.cpp (hunk 1 above)
Comment (jmosborne) [reply mesh/src/mutable/Cylindrical2dMesh.cpp]: Fixed and updated a few more I caught in the classes.

Comment (jmosborne) [reply mesh/src/mutable/Cylindrical2dMesh.cpp]: Fixed this and others I found.

File: mesh/src/vertex/Cylindrical2dVertexMesh.cpp (hunk 4 above)
Comment (jmosborne) [reply mesh/src/vertex/Cylindrical2dVertexMesh.cpp]: Done along with rest of the occurrences in the file.

File: mesh/src/mutable/Cylindrical2dMesh.cpp (hunk 3 above)
Comment (jmosborne) [reply mesh/src/mutable/Cylindrical2dMesh.cpp]: Done

File: mesh/src/mutable/Cylindrical2dMesh.cpp (hunk 2 above)
Comment (jmosborne) [reply mesh/src/mutable/Cylindrical2dMesh.cpp]: Done

File: mesh/src/vertex/Cylindrical2dVertexMesh.cpp (hunk 5 above)
Comment (jmosborne) [reply mesh/src/vertex/Cylindrical2dVertexMesh.cpp]: Removed

File: mesh/src/vertex/Cylindrical2dVertexMesh.cpp (hunk 6 above)
Comment (jmosborne) [reply mesh/src/vertex/Cylindrical2dVertexMesh.cpp]: uncommented it as should be uncommented

File: mesh/src/vertex/Cylindrical2dVertexMesh.cpp (hunk 8 above)
Comment (jmosborne) [reply mesh/src/vertex/Cylindrical2dVertexMesh.cpp:428]: The code is 

if (x_location < 0)
    {
        pNode->rGetModifiableLocation()[0] = x_location + mWidth;
    }
    else if (x_location >= mWidth)
    {
        pNode->rGetModifiableLocation()[0] = x_location - mWidth;
    }

So if the second one is < then modes  at mWidth aren't moved to 0 which isn't correct as as here 0=mWidth

Ive updated the hpp to have the correct closed/open interval [0,mWidth)

File: mesh/test/mutable/TestCylindrical2dMesh.hpp (hunk 9 above)
Comment (jmosborne) [reply mesh/test/mutable/TestCylindrical2dMesh.hpp:329]: Fixed

**Comment by jmosborne** [comment]
Issue for this PR is https://github.com/Chaste/Chaste/issues/142.

I fixed the comments. I fixed most formatiing errors manually as when i did it automatically 50% of lines in the files had changes. all formatting is correct in new code (and some old code) so its a net positive change!

File: mesh/src/vertex/Cylindrical2dVertexMesh.cpp (hunk 7 above)
Comment (jmosborne) [reply mesh/src/vertex/Cylindrical2dVertexMesh.cpp]: I have reworded this and another todo and left as a future change as these chosen varies are sufficient for the current use cases.

**Review by fcooper8472 — APPROVED** [review]

**Pull request merged**